### ISSUE TITLE
Provide clone methods for a variety of objects

### ISSUE THREAD

**Issue by wnbittle** [dyn4j/dyn4j#293]
A user may need to clone a Body or a Shape at any time during their program.  Cloning today could be challenging as some class members are final, protected, or private.

This enhancement proposes adding the Clonable<T> interface to a variety of classes, including:

- AbstractCollisionBody
- AbstractPhysicsBody
- Shape

A full analysis is needed to determine the scope of the change, but I'm not seeing any blockers to this feature.

**Comment by JNightRide** [comment]
Hi @wnbittle 

In relation to support for cloning and what is mentioned in this [PR](https://github.com/dyn4j/dyn4j/pull/292); It is possible to clone such an object with some "tricks"... I haven't submitted any PR at this time, however I would like you to check out this [branch](https://github.com/JNightRide/dyn4j/tree/cloning).

In summary, class `Copier` is responsible for cloning the object + the fields of said clone to have a complete object with all the characteristics, for the moment; only the 'Capsule' class supports cloning (`BodyFixure`)...

What do you think of this idea? At the moment it is a draft and it is possible to improve it to adapt it to the different needs that arise.

If you agree, please submit a PR with these changes (subject to change)

[**NOTE**]
There is a (rather ugly) unit test that demonstrates cloning: [CloneTest](https://github.com/JNightRide/dyn4j/blob/cloning/src/test/java/org/dyn4j/simulation/CloneTest.java)

**Comment by wnbittle** [comment]
TLDR; I'm still evaluating options and trying things.  One more question for you @JNightRide could you elaborate more on _why_ you want to clone a body?

### Copying / Cloning

I've been messing around with a few strategies.  I'm recording a few gotchas that I've found so far with copy/clone:

1) The `Copyable` interface isn't enough for full typesafety:
```java
public AbstractCollisionBody(AbstractCollisionBody<T> source) {
	this();
	for (int i = 0; i < source.fixtures.size(); i++) {
		T fix = source.fixtures.get(i);
		T n = (T)fix.copy();
		this.fixtures.add(n);
	}
}
```
In this snippet, the key issue is `T n = (T)fix.copy();`.  We know that `T` is something that extends `Fixture`.  `Fixture` implements the `Copyable<Fixture>` interface.  What this means is that the `copy` method on `T` here returns `Fixture` not `T`.  Casting it fixes the problem, but it assumes that the `copy` method is overridden in every subclass of `Fixture`.

2) The `Link` class is a real pain.  Imagine the following code:
```java
List<Link> links1 = Geometry.createLinks(...);
List<Link> links2 = Geometry.createLinks(...);
Body b1 = ...;
for (Link link : links1) {
  b1.addFixture(link, ...);
}
for (Link link : links2) {
  b1.addFixture(link, ...);
}
```
In this snippet a user has created two `Link` chains and added them to the same `Body`.  If I wanted to copy that `Body`, I would copy each `Fixture` and therefore each `Link`.  The problem is that they can't be copied one-by-one like everything else would be because they hold next/previous references.  Instead, there has to be a specific process to handle `Link`s something like `if (fixture.getShape() instanceof Link)`.  I can't express how much I don't like that.

3) The semantics for a clone/copy operation are somewhat different for each context.  For `BodyFixture`, `Shape`, and other primitives, it's easy, they are 1 for 1 copies - completely identical (with exception of `userData`).  But for `CollisionBody` and `PhysicsBody` there are some fields we shouldn't copy.  For `Joint`s it's different again, you wouldn't want to duplicate a joint for the same two bodies - that doesn't do anything but over constrain the simulation, instead you'd probably want to clone the Joints and associate it to cloned bodies - but does that mean you would clone the `Joint` first or would cloning a `Joint` require you to supply the bodies as arguments?

### Serialization / Deserialization
I was messing with serialization at the same time since it's very similar to object cloning.  For example consider this:
```java
Object obj = ...;
byte[] data = obj.serialize();
Object obj2 = data.deserialize();
```
`obj2` is a copy of `obj` just by nature of serializing and deserializing.  As I've went down this rabbit hole, I've had a few discoveries:

1) JoltPhysics / Bullet take the approach of writing the serializable bits to a `byte[]`, something like this:
```java
public void serialize(Stream stream) {
  stream.writeDouble(...);
  stream.writeFloat(...);
  //...
}
```
What's nice about this is that the data is packed well and there's no new data-only classes used.  What's not so great about this is that it assumes the field order and doesn't leave a lot of flexibility as the classes evolve over time.

2) I tried using specific serialization classes for each class, like this:
```java
public class CircleData {
  private double radius;
}
public class Circle {
  public CircleData save() {
    CircleData data = new CircleData();
    data.radius = this.radius;
    return data;
  }
}
```
The challenge I had with this was yet again inheritance related.  I can't define a `Saveable<T>` interface and apply it to the `Shape` interface because the contract would return a `ShapeData` (even if `Circle` returns `CircleData`).  This is the same problem as the copy issue above.  One thing I really don't like about this is maintain a whole new set of classes just for serialization that should have the same fields as their companion class - it seems like overkill.

3) I tried using a generic serialization class for all classes in a hybrid like approach.  It went something like this:
```java
public class Data extends HashMap<String, Object> {}

public class Circle {
  public Data save() {
    Data data = new Data();
    data.put("radius", this.radius);
    return data;
  }
}
```
So the biggest issue here is the hard-coded key names.  On one hand it's nicer because it's very explicit but it's not as rigid as the previous option.  

4) Where all these fall flat on their face is during deserialization.
```java
Body b1 = ...;
b1.addFixture(Geometry.createCircle(1.0));
b1.addFixture(Geometry.createRectangle(1.0, 2.0));
Data data = b1.save();
// this is where the problem is
Body b2 = new Body(data);

// meanwhile in Fixture
public class Fixture {
  public Fixture(Data data) {
    // how do I know what shape to create here?
    // it will probably devolve into something nasty like this:
    // find the class by name (assuming it was stored in Data)
    Class<?> clazz = Class.forName(data.getType());
    // now find the constructor that accepts a Data type
    // then call that constructor with the data and cast it
    Convex shape = (Convex)clazz.getConstructor(Data.class).newInstance(data.get("shape", Data.class));
  }
}
```
This is just really nasty.  Surprisingly though, this is par for the course for this kind of stuff.  Bullet and JoltPhysics do it this way (except they use enums/constants and class registering) and this is what I've done in the past on other projects.

5) Serialization really needs stable identifiers for things that get serialized for two reasons.  One, they are used to solve the saving of graph structures where just following reference chains would loop infinitely.  Two, you don't want to generate a different output every time you run the serializer.  If I run the serializer on body 1, I should get the same ID for it every time.  My thought here was something like this:
```java
public final class IdGenerator {
  private static final AtomicLong index = new AtomicLong(0);
  public static long getNextId() {
    return index.incrementAndGet();
  }
}
```
This approach allows generating ids easily, but it would not be safe for continuing the id counting after loading a simulation.  Any other context-based approach would require a new constructor argument and break compatibility.

**Comment by JNightRide** [comment]
> One more question for you @JNightRide could you elaborate more on why you want to clone a body?

I need to clone the physical bodies for fast resource loading; I am currently using JME3 to create 2D games, said engine has an asset loader that uses a cache system to load resources quickly; However, to have it it is necessary that the objects support cloning (fast loading).

These resources I'm trying to load are serialized (includes dyn4j physical bodies), I just want a copy of the object that is in memory without affecting it....

 > For Joints it's different again, you wouldn't want to duplicate a joint for the same two bodies - that doesn't do anything but over constrain the simulation, instead you'd probably want to clone the Joints and associate it to cloned bodies - but does that mean you would clone the Joint first or would cloning a Joint require you to supply the bodies as arguments?

I believe that the cloning of a Joint does not necessarily have to be associated with previously cloned bodies.

```java
Body body = new Body();
Body clonBody = body.copy(); // This body is independent.

PinJoint<Body> joint = new PinJoint<Body>(body, ...);
                
// This 'oint' is independent of the previously
// cloned physical body
PinJoint<Body> clonJoint = joint.copy();
Body jointBody = clonJoint.getBody();
```

If I want a copy, it wouldn't make sense to manually clone the physical bodies (user); It is better to clone the joint along with the bodies and individually access the bodies of that new joint.

**Serialización/Deserialización**
> JoltPhysics / Bullet take the approach of writing the serializable bits to a byte[], something like this
...
What's nice about this is that the data is packed well and there's no new data-only classes used. What's not so great about this is that it assumes the field order and doesn't leave a lot of flexibility as the classes evolve over time.

I'm not very familiar with the JoltPhysics / Bullet serialization process. From what you can see, it's not too far from the process that JME uses... chances are we can implement a system similar to them.

> This is just really nasty. Surprisingly though, this is par for the course for this kind of stuff

Certainly working with metaprogramming is somewhat tedious and sometimes a bit complicated due to the different accesses.

With serialization/deserialization; Wouldn't it be great to move this topic to a new "issue"?

## Copyable

Returning to the topic of cloning, I think that implementing an intermediary object that is responsible for managing clones can solve the problems of cloning 'Link' objects; This object will save information (reference) of the cloned object in memory, when the same object is requested to be cloned, the previously cloned object (memory) is returned.

```java
public AbstractCollisionBody(AbstractCollisionBody<T> source) {
	this();                
    final Copier copier = new Copier(); // clone manager
     this.transform.set(source.getTransform().copy());
     for (final Fixture fixture : source.fixtures) {
         T fix = copier.copy(fixture);
         this.fixtures.add(fix);
     }                
     if (copier.isSupported(source.userData)) {
         this.userData = copier.copy(source.userData);
     }                
     this.radius = source.radius;
     this.enabled = source.enabled;
}
```

This would update the `Copyable` interface:
```java
public interface Copyable<T extends Copyable<? extends T>> {
	public T copy();
        public void copyFields(Copier copier, T original);
}

// class - 'Link'
public class Link extends Segment implements ...Copyable<Link> {
	...       
        private Link() {
            this(new Vector2(0, 0), new Vector2(1, 1));
        }

        @Override
        public Link copy() {
            return new Link();
        }

        @Override
        public void copyFields(Copier copier, Link original) {
            this.previous = copier.copy(original.previous);
            this.next = copier.copy(original.next);
        }
        ...
}

// class - Copier
public final class Copier {
    
        private final IdentityHashMap<Object, Object> cache = new IdentityHashMap<>();
    
        public boolean isSupported(Object object) {
                return (object instanceof Copyable);
        }
    
        public <T> T copy(T object) {
                if (object == null) {
                        return null;
                }
                
                Class<T> type = (Class<T>) object.getClass();
                Object copy = cache.get(object);
                if (copy != null || cache.containsKey(object)) {
                        return type.cast(copy);
                }
                
                if (object instanceof Copyable) {
                        copy = ((Copyable) object).copy();
                        
                        cache.put(object, copy);
                        
                        ((Copyable) copy).copyFields(this, (Copyable) object);
                } else {
                         throw new IllegalArgumentException("Object is not cloneable, type:" + type);
                }
                return type.cast(copy);                
        }
}
```

What do you think of this idea?

**Comment by wnbittle** [comment]
@JNightRide Would you be able to list all the objects you want copyable?  You've focused on `Body`, but I'm assuming there are others.

**Comment by JNightRide** [comment]
@wnbittle
First of all, thanks for everything (for support of cloning).

> Would you be able to list all the objects you want copyable?

1. The first are the physical bodies (already mentioned)
2. The different shapes (`Circle, Rectangle...`)
3. `Fixture` -> `BodyFixture`
4. `Force` (is easy to copy but preferably provides a method to copy it)
5. `Torque` (is easy to copy but preferably provides a method to copy it)
6. Copying the `Joins` can be a little difficult, but it would be great to be able to copy them.
7. `Filter`
    - As for filters, I think only one class supports cloning (`CategoryFilter`)

I can't think of any other object.

**Comment by wnbittle** [comment]
Thank you for being patient and for the list.  I am working on it, but I'm debating some hard decisions that will break backward compatibility, but will be better for copy/serialization in the long run.

A few more questions, just to make sure I'm solving the need appropriately:

1. Are you looking for the native `clone` method to be supported or would supporting a custom `copy` method work?  I don't want to support `clone` for a number of reasons, some philosophical, some practical.
2. When I'm thinking copy, I'm thinking a deep copy, nothing on the copy references the original object - you are expecting the same right?

**Comment by JNightRide** [comment]
>Are you looking for the native clone method to be supported or would supporting a custom copy method work? I don't want to support clone for a number of reasons, some philosophical, some practical.

a native clone (`clone()`) doesn't really matter, a custom method is a very good idea considering **dyn4j** is a physics engine.

> When I'm thinking copy, I'm thinking a deep copy, nothing on the copy references the original object - you are expecting the same right?

exactly, a copy that has no connection to the original object.

**Comment by wnbittle** [comment]
Perfect, thank you for confirming those assumptions.

**Comment by wnbittle** [comment]
@JNightRider I've checked in the changes for [copy support](https://github.com/dyn4j/dyn4j/commit/ef3a0b8a3c0ec504b3b7bbbc0c46a03d5e714dd9) and I'm mostly happy with it.  I may change one or two things before release.  I'd love it if you'd review it and make sure it meets your needs.

Another question - I've been playing around with native serialization support but it's going very slow - do you need serialization or would you be unblocked if I released the next version with only copy support?

**Comment by JNightRider** [comment]
> I've checked in the changes for [copy support](https://github.com/dyn4j/dyn4j/commit/ef3a0b8a3c0ec504b3b7bbbc0c46a03d5e714dd9) and I'm mostly hppy with it

GREAT, THANKS FOR SUPPORTING CLONING!

> I'd love it if you'd review it and make sure it meets your needs.

I will start doing rigorous testing, these tests may take a while as dyn4j has changed a lot with this new version (I'm excited to try it out)

> Another question - I've been playing around with native serialization support but it's going very slow - do you need serialization or would you be unblocked if I released the next version with only copy support?

I guess the build time is a bit long (slow). As for serialization, publishing it in the future (separately) wouldn't affect anything. I think clone support (without serialization) is more than enough for me (for now)...

**Comment by JNightRider** [comment]
Hi @wnbittle 

I was doing some tests in general, all the bodies and arciculations were copied efficiently (it's very good).

Except for one small problem (I don't know if it really is), when I create a list of 'Links' without closing it to clone them manually, it causes NPE:

```java
        List<Link> links = Geometry.createLinks(new Vector2[]{
                new Vector2(0,0), new Vector2(1, 1),
                new Vector2(2, 2), new Vector2(3, 3)
        }, false);

        List<Link> copia = new ArrayList<>();
        for (final Link l : links) {
            copia.add(l.copy());
        }
```

It seems that vector 0 of some link is null and dyn4j is trying to forcefully copy it.

[ error output ]

```shell
Exception in thread "main" java.lang.NullPointerException: Cannot invoke "org.dyn4j.geometry.Vector2.copy()" because "link.point0" is null
	at org.dyn4j/org.dyn4j.geometry.Link.<init>(Link.java:73)
	at org.dyn4j/org.dyn4j.geometry.Link.copy(Link.java:84)
	...
```

**Comment by wnbittle** [comment]
Ahh great catch!  I'll get an update committed today to fix along with a unit test.

**Comment by wnbittle** [comment]
@JNightRider  I've fixed the issue with the `Link` copy.  If you wouldn't mind, please run another test to verify.

**Comment by JNightRider** [comment]
I've been using the new version in my projects (with the latest part released), everything works as usual; There is no regression at the engine level.

Something curious that I found at the code level, in package `org.dyn4j.world.result` All objects seem copyable, said copy is superficial - Is it normal for this to be the case? Would it affect whether a full copy is applied?

It originates from class `DetectResult,` everything passes through the constructor by reference without copying the objects themselves.

```java
...
public class DetectResult<T extends CollisionBody<E>, E extends Fixture> implements Copyable<DetectResult<T, E>> {
	...
	protected DetectResult(T body, E fixture) {
		this.body = body;
		this.fixture = fixture;
	}
        ...
        public DetectResult<T, E> copy() {
		return new DetectResult<T, E>(this.body, this.fixture);
	}
}
```

This seems to have been around since version 5.0 (maybe a few versions earlier), I assume a shallow copy is used because physical bodies didn't support copying...
